File: blind75/arrays&Hashing/49-groupAnagrams.js
// 49. Group Anagrams
// Given an array of strings strs, group the anagrams together. You can return the answer in any order.
// An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
    let aMap = {};

    for (let str of strs) {
        let arr = new Array(26).fill(0);

        for (let i = 0; i < str.length; i++) {
            let index = str.charCodeAt(i) - "a".charCodeAt(0);
            arr[index]++;
        }

        let key = arr.toString();
        if (aMap[key] === undefined) {
            aMap[key] = []
        }
        aMap[key].push(str)
    } 
    
    return Object.values(aMap);
};

// 这道题 用一个array，array[0-25] 记录着从'a'-'z'的出现次数
// 再把这个array转换为string
// 这样的话 如果是anagram，key应该相同
// 如果key相同就把value记录下来


let strs = ["eat","tea","tan","ate","nat","bat"];
let strs1 = [""];
let strs2 = ["a"];

console.log(groupAnagrams(strs));
console.log(groupAnagrams(strs1));
console.log(groupAnagrams(strs2));

File: neetcodeAll/array&Hashing/49-groupAnagrams.py
# 49. Group Anagrams
# Given an array of strings strs, group the anagrams together. You can return the answer in any order.
# An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

from typing import List
from collections import defaultdict

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        res = defaultdict(list)
        for str in strs:
            key = "".join(sorted(str))
            res[key].append(str)
        return list(res.values())

# 我们可以遍历这个strs，然后以它的sorted值为key，他自己为value

strs = ["eat","tea","tan","ate","nat","bat"]
print(Solution().groupAnagrams(strs))
strs = [""]
print(Solution().groupAnagrams(strs))
strs = ["a"]
print(Solution().groupAnagrams(strs))
